indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/common/OrderedPartitionableRecord.java

```diff
@@ -74,7 +74,7 @@ public SequenceOffsetType getSequenceNumber()
   }
 
   @NotNull
-  public List<byte[]> getData()
+  public List<? extends ByteEntity> getData()
```

Comment:
This change means we're returning a `ByteBuffer` instead of a `byte[]`, which opens up potential issues due to callers handling the position and limit wrong. What are the expectations? Is it ok for callers to change the position of the underlying buffer or should they refrain? If they do change it, does that cause problems?

(For example: do we have situations where something calls `getData()` to parse a value, and that updates the position, and then later on `getData()` is called again for some other purpose like logging? If so — that'd break as a result of this change.)

Whatever the case, the javadoc for this method should describe the conclusion.

Reply:
I did encounter some of this in test cases, where we would call getData on a record multiple times, causing issues. In general however, an InputEntity doesn't support being read multiple times, e.g. for `ByteEntity`, once it's been read via `InputEntity.open()`, the underlying buffer would have been exhausted and return an empty stream on subsequent calls to `open()`.

Except for one logging case, all the calls to `OrderedPartitionableRecord.getData()` I could find would only be executed once per record, for the purpose of exposing InputEntities to parsers, which should ensure the buffers only get read once or treated appropriately for cases. I updated the logging case to make sure it wouldn't modify the buffer position.